[distributed] add timeout parameter to connect function in TCPStore

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26554 [distributed] add timeout parameter to connect function in TCPStore**

Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

rohan-varma committed:
[pytorch][distributed] add timeout parameter to connect function in TCPStore

Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout parameter to connect function in TCPStore"


Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

pietern commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -286,8 +287,10 @@ TCPStore::TCPStore(
         new TCPStoreDaemon(masterListenSocket_));
   }
+  // set the timeout
+  setTimeout(timeout);
```
Ideally this is set directly in the base class constructor. Can you add a constructor to `Store.hpp` that takes the timeout and sets it? We can keep the existing default constructor that initializes with the default timeout.

pietern commented on torch/lib/c10d/TCPStore.cpp:
```diff
@@ -286,8 +287,10 @@ TCPStore::TCPStore(
+  setTimeout(timeout);
   // Connect to the daemon
-  storeSocket_ = tcputil::connect(tcpStoreAddr_, tcpStorePort_);
+  storeSocket_ = tcputil::connect(tcpStoreAddr_, tcpStorePort_, true, timeout_);
```
Please add `/* wait= */` before `true` so you can read immediately what it is for. The other args are quite clear except for the wait one. This is useful for opaque params where you cannot infer what's happening by looking at it.

pietern reviewed (commented):
Thanks, @rohan-varma!

rohan-varma committed:
Update on "[distributed] add timeout parameter to connect function in TCPStore"


Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[distributed] add timeout parameter to connect function in TCPStore"


Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

pietern reviewed (approved):
Looking good. Thanks, @rohan-varma!

pietern commented:
Please wait for CI to complete before merging.

rohan-varma committed:
Update on "[distributed] add timeout parameter to connect function in TCPStore"


Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma commented:
@pytorchbot retest this please

rohan-varma committed:
Update on "[distributed] add timeout parameter to connect function in TCPStore"


Previously, in `TCPStore`'s constructor we did not pass in a timeout to
the `connect` function, which thus used the default timeout (-1, so infinite).
But the timeout variable in `TCPStore.cpp `is configurable by the user and set to
be 300 seconds by default, so we should be passing this into the connect function. (See https://github.com/pytorch/pytorch/issues/26513)

Test plan: force connection errors and verify that the timeout is respected:
1) Without applying this change, force connection error by setting `tcpStorePort_=1` in `TCPStore.cpp`
2) Build the `TCPStoreTest` binary.
3) Run the binary. Observe that `connect()` does not return.
4) Apply this change. Observe that `connect()` does time out.

Differential Revision: [D17486779](https://our.internmc.facebook.com/intern/diff/D17486779/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot retest this please

rohan-varma commented:
@pytorchbot retest this please

rohan-varma commented:
@pytorchbot retest this please

pietern commented:
@rohan-varma I don't think this will pass again. The error seems to indicate that the CircleCI workspace for this job (the root job that kicked off all the builds for this commit) has gone. The only way to retrigger is to push a new commit, I think (can you confirm, @ezyang?).

ezyang commented:
Yes

pietern commented:
@ezyang TYVM!
